### PR TITLE
stdenv: enable __structuredAttrs

### PR DESCRIPTION
**Pull request by globin**
`__structuredAttrs` is a nice feature, we want it enabled in stdenv!

hydra jobset: https://hydra.nixos.org/jobset/nixpkgs/structured-attrs

related: nixos/rfcs#13

Further information: https://nixos.mayflower.consulting/blog/2020/01/20/structured-attrs/

cc @Ericson2314 @matthewbauer

### CODE REVIEW

File: pkgs/development/compilers/gcc/7/default.nix [hunk 1]
```diff
@@ -335,21 +335,21 @@ stdenv.mkDerivation ({
...
   # LIBRARY_PATH= makes gcc read the specs from ., and the build breaks.
 
-  CPATH = optionals (targetPlatform == hostPlatform) (makeSearchPathOutput "dev" "include" ([]
+  CPATH = toString (optionals (targetPlatform == hostPlatform) (makeSearchPathOutput "dev" "include" ([]
     ++ optional (zlib != null) zlib
-  ));
+  )));
 
-  LIBRARY_PATH = optionals (targetPlatform == hostPlatform) (makeLibraryPath (optional (zlib != null) zlib));
+  LIBRARY_PATH = toString (optionals (targetPlatform == hostPlatform) (makeLibraryPath (optional (zlib != null) zlib)));
```
Comment (volth) [inline pkgs/development/compilers/gcc/7/default.nix]: `makeLibraryPath` returns string, so why `toString` and not just
```diff
- LIBRARY_PATH = toString (optionals (targetPlatform == hostPlatform) (makeLibraryPath (optional (zlib != null) zlib)));
+ LIBRARY_PATH = optionalString (targetPlatform == hostPlatform) (makeLibraryPath (optional (zlib != null) zlib));
```

File: pkgs/development/libraries/glibc/common.nix [hunk 2]
```diff
@@ -186,7 +187,7 @@ stdenv.mkDerivation ({
     configureScript="`pwd`/../$sourceRoot/configure"
 
     ${lib.optionalString (stdenv.cc.libc != null)
-      ''makeFlags="$makeFlags BUILD_LDFLAGS=-Wl,-rpath,${stdenv.cc.libc}/lib"''
+      ''makeFlags+=("BUILD_LDFLAGS=-Wl,-rpath,${stdenv.cc.libc}/lib")''
```
Comment (volth) [inline pkgs/development/libraries/glibc/common.nix]: `makeFlags` -> `makeFlagsArray` ?

Comment (globin) [reply pkgs/development/libraries/glibc/common.nix]: makeFlags is an array now too

makeFlagsArray will be deprecated as it is of no additional use with structuredAttrs

**Comment by jtojnar** [comment]
Why not make the `NIX_*_FLAGS` be lists as well?

**Comment by globin** [comment]
It needs to be exported to be picked up by bintools-/cc-wrapper

**Comment by Ericson2314** [comment]
@globin we could require that they be lists, but then convert to strings (and quote!) ourselves. If that is not more churn, I think i might prefer that.

**Comment by Ericson2314** [comment]
I added `set -u` as  I got that ready/pretty close, and ensured at least `stdenv` built after the merge. Also fixed an eval failure.

File: pkgs/stdenv/generic/setup.sh [hunk 3]
```diff
@@ -827,7 +827,7 @@ unpackFile() {
 unpackPhase() {
     runHook preUnpack
 
-    if [ -z ${srcs+"${srcs[@]}"} ]; then
+    if [ -z ${srcs-"${srcs[@]}"} ]; then
```
Comment (Ericson2314) [inline pkgs/stdenv/generic/setup.sh]: Intentional?

**Comment by Ericson2314** [comment]
I added another big round of `set -u` fixes. Unfortunately I accidentally added a tab before, and fixing that made it a full stdenv rebuild.

File: pkgs/development/libraries/gstreamer/base/default.nix [hunk 4]
```diff
@@ -80,7 +80,7 @@ stdenv.mkDerivation rec {
     ++ lib.optional (!stdenv.isDarwin) libvisual
     ++ lib.optional enableAlsa alsaLib
     ++ lib.optionals enableX11 [ libXv pango ]
-    ++ lib.optional enableWayland wayland
+    ++ lib.optionals enableWayland [ wayland wayland-protocols ]
```
Comment (Ericson2314) [inline pkgs/development/libraries/gstreamer/base/default.nix]: Is perhaps now needing wayland-protocols symptomatic of an underlying issue that we should fix instead?

Comment (globin) [reply pkgs/development/libraries/gstreamer/base/default.nix]: Yeah, this looks like it should be fixed at some other point, although I cannot see at first glance what might be causing it.

**Comment by asymmetric** [comment]
[More info](https://github.com/NixOS/nix/commit/6de33a9c675b187437a2e1abbcb290981a89ecb1) about `__structuredAttrs`, for those who are wondering.

**Comment by timbertson** [comment]
If anyone else is wondering how this actually hangs together after reading the nix PR (which only talks about a JSON file, not something that bash supports terribly well: nix also produces an `.attrs.sh`, which exposes the same JSON in a bash-y way (added in https://github.com/NixOS/nix/commit/2d5b1b24bf70a498e4c0b378704cfdb6471cc699). Many of the diffs here source that file, but it was unclear to me where it came from.

File: pkgs/development/compilers/gcc/8/default.nix [hunk 5]
```diff
@@ -223,56 +220,19 @@ stdenv.mkDerivation ({
...
-  ${if hostPlatform.system == "x86_64-solaris" then "CC" else null} = "gcc -m64";
-
-  # Setting $CPATH and $LIBRARY_PATH to make sure both `gcc' and `xgcc' find the
-  # library headers and binaries, regarless of the language being compiled.
-  #
-  # Likewise, the LTO code doesn't find zlib.
-  #
-  # Cross-compiling, we need gcc not to read ./specs in order to build the g++
-  # compiler (after the specs for the cross-gcc are created). Having
-  # LIBRARY_PATH= makes gcc read the specs from ., and the build breaks.
```
Comment (Ericson2314) [inline pkgs/development/compilers/gcc/8/default.nix]: I think this stuff was deleted by mistake? I'll take a look

Comment (globin) [reply pkgs/development/compilers/gcc/8/default.nix]: should be in the `env` attrs

Comment (Ericson2314) [reply pkgs/development/compilers/gcc/8/default.nix]: Oh, OK.

**Comment by jtojnar** [comment]
Would it make sense to cherry pick trivial fixes that work on master like https://github.com/NixOS/nixpkgs/commit/3cc957834bb4cd0626f2cf680e525eedbdf9e8f5 into master so that this easier to review?

Maybe even temporarily change `mkDerivation` on master to merge `env` into the derivation and cherry-pick the `env` fixes too.

And then rebase this onto master.

**Comment by Ericson2314** [comment]
I think we were trying to cherry-pick in the non-breaking changes and then merge back into master already, but yes, we will definitely need to do more of that to get this over the finish line.

**Comment by jtojnar** [comment]
I am going to rebase this then.

**Comment by Ericson2314** [comment]
I think we merged master into it before, but fine either way, whatever you prefer.

**Comment by jtojnar** [comment]
I have cherry picked most of current-stdenv-compatible changes into staging, fixing some minor typos and oversights and hopefully not introducing to many new ones. What should remain here, for the most part, are the trivial `env.` additions, and the changes to stdenv itself.

It would be great if we could merge a trivial env shim into `mkDerivation` on master that would allow us to move the `env.` changes out of this PR and focus solely on the breakage. Not sure what should happen when the same attr is passed to both `mkDerivation` directly and through `env`, though.

It might also be interesting to add some kind of structured counterpart to `env.NIX_CFLAGS_COMPILE` etc. But not sure if it is worth the effort.

Also, I do not understand the [current error](https://gist.github.com/GrahamcOfBorg/27654479a78d4bc9096b438c5cfccc26) – why would `builtins.isPath` complain about “attribute 'isPath' missing”?

**Comment by jtojnar** [comment]
sudo flags were [reverted](https://github.com/NixOS/nixpkgs/pull/76705/commits/ee3f276461414d96830a4a84152b259536478ce3) since spaces in `*Flags` are problematic on master:

```
install flags: SHELL=/nix/store/j78ycszn9vgn0khl0lcpbd087x2mmr66-bash-4.4-p23/bin/bash install_uid=1000 install_gid=100 sudoers_uid=\$\(shell id -u\) sudoers_gid=\$\(shell id -g\) sysconfdir=/nix/store/3w2rk7pf2cbvgpl901z8yjiv6sdqqqay-sudo-1.8.29/etc rundir=\$\(TMPDIR\)/dummy vardir=\$\(TMPDIR\)/dummy DESTDIR=/ install
make: invalid option -- 'u'
make: invalid option -- ')'
make: invalid option -- 'g'
make: invalid option -- ')'
Usage: make [options] [target] ...
```

**Comment by globin** [comment]
I just published a blog post with some reasoning on why we would want `__structuredAttrs` and why which changes are needed: https://nixos.mayflower.consulting/blog/2020/01/20/structured-attrs/

**Comment by Ericson2314** [comment]
@teto Good question. I am not sure why the number of commits and diff is so large at the top, especially when @jtojnar was merging in bits of this that were ready. Maybe github is wrong again? My guess is that it would be best to 

1. Ensure thing are merged in when possible so this doesn't grow faster than it needs to.

2. Try to get things at least evaluating before worrying about getting them building.

But maybe there's another strategy here I am missing.

**Comment by Ma27** [comment]
> Maybe github is wrong again? My guess is that it would be best to

It's because master has been merged into this branch, however the PR points to staging I guess.

@teto,

Evaluation errors etc can be found at the corresponding hydra jobset: https://hydra.nixos.org/jobset/nixpkgs/structured-attrs

If you have a google-account you may login there to trigger evals. If you want to fix builds as well, you may want to check the delta of failing packages between trunk and structured attrs (this can be done using the `Compare to...` button: https://hydra.nixos.org/eval/1579112?compare=trunk)

**Comment by bjornfor** [comment]
> But maybe there's another strategy here I am missing.

How about having a the new structured stdenv as `stdenv2`, while packages get migrated? And finally rename stdenv2 to stdenv? If this is doing everything in-place I'm afraid of breakage.

**Comment by teto** [comment]
@Ma27 thanks for the useful tips. I don't have a google account there but I have a server where I can build stuff. How can I see what hydra's eval look like ? is that ok `nix-instantiate ./pkgs/top-level/release.nix --show-trace --option restrict-eval true -I foo=. --arg nixpkgs '{ outPath = ./.; revCount = 0; shortRev = "aabbcc"; }` ?

I would like to push fixes to this branch but in order to not duplicate work, I would like to announce them first and then push (and confirm others look at the list/do the same). 
Once we fix eval, I wonder if it would be fine to just merge and pause "nixos-unstable" for a while as we fix the builds.

**Comment by Ma27** [comment]
> How can I see what hydra's eval look like

At the jobset view there's an `Evaluation errors`-tab (https://hydra.nixos.org/jobset/nixpkgs/structured-attrs#tabs-errors)

> I would like to push fixes to this branch but in order to not duplicate work, I would like to announce them first and then push (and confirm others look at the list/do the same).

Sounds reasonable. But please note that I'm not actively working on this currently. You should probably talk to @lheckemann and @globin in this case.

**Comment by Ericson2314** [comment]
@FRidh well it doesn't support lists because it doesn't want to automatically concat them into env vars, as bash arrays cannot be exported. (If you try, you just export the first element, I believe.)

**Comment by globin** [comment]
I have purposefully avoided that, I'd prefer keeping env.* vars consistently scalar types. If necessary people can themselves build lists and `toString` it before passing it to the `env` attrs. They then should be aware that there is no special handling of arg splitting etc in `env`.

**Comment by teto** [comment]
Hydra is triggered manually for good reasons but would it be possible to configure hydra to rebuild automatically some basic packages like gcc ? (and also disabling the darwin platform since it doesn't help much in finding bugs, rather the contrary ). I would like to fix the build of for instance qt4 (I got a local fix) but having to rebuild gcc is too much.

**Comment by Ma27** [comment]
`stripDebugList` has been fixed in #84551

**Comment by teto** [comment]
We've just had the release so maybe we could make some plan to enable this in the next one ? have people coordinate (assign pkgs/applications/video/ to X etc) and rebase on top of nixos-unstable regularly ? like a nixos-unstable-attrs channel

**Comment by rnhmjoj** [comment]
Has the work to enable structured attributes stalled os is progressing somewhere else?
It seems to me this PR is too big to review and keep in sync with master.
Is it possible to merge this progressively into master without breackages?

**Comment by xaverdh** [comment]
I think #76732 and #85042 are more viable approaches. Both are somewhat stalled atm though.

**Comment by vcunat** [comment]
I assume this PR is complex enough to be considered a "breaking change", right?  Therefore not fitting [21.05 schedule](https://discourse.nixos.org/t/21-05-release-schedule/12528) anymore.

File: pkgs/stdenv/generic/setup.sh (hunk 3 above)
Comment (andersk) [reply pkgs/stdenv/generic/setup.sh:854]: `${srcs-"${srcs[@]}"}` expands to the first element of the `srcs` array; surely that’s not the intention.

```suggestion
    if [ ${#srcs[@]} = 0 ]; then
```

**Comment by Artturin** [comment]
A way to help could be:
fork nixpkgs and checkout the structured-attrs branch
run `git rebase -i origin/master` and solve the merge conflicts (300 or so)

**Comment by Artturin** [comment]
what if we created a `structuredAttrsStdenv` and migrated expressions gradually?

**Comment by Mindavi** [comment]
At least making it so this change can be split up seems like a good idea to me. I don't mind hacking at this, but doing everything at once is very hard to do, it'll also quickly have conflicts.

**Comment by Artturin** [comment]
@Mindavi I found https://github.com/NixOS/nixpkgs/pull/85042
Which has structuredAttrs as a opt in

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/175649

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/175649 has been merged

**Pull request closed without merging**